### PR TITLE
Devices and external libraries options, make install, Makefile fix 

### PR DESCRIPTION
**Pull request by nicolaspi**
- New features that allow to generate code according to operator's devices compatibility (cpu and/or gpu), and to specify external libraries to be linked.
- The Makefile has been fixed to work with the latest tensorflow version (tested on the version 1.3 built from source).
- Added 'make install' that copies the built library to the tensorflow's library directory.

### CODE REVIEW

**Comment by nicolaspi** [comment]
Looks like a version issue with tensorflow, the linker does not find  `libtensorflow_framework.so` which was introduced by [this](https://github.com/tensorflow/tensorflow/commit/5c7f9e316d8c7735308a217310350d416d7498cc) commit (19 days ago).

**Comment by sjperkins** [comment]
Thanks for the PR. It looks like the  new `libtensorflow_framework.so` linking style is going into tensorflow 1.4.0 (see https://github.com/tensorflow/tensorflow/issues/13607 for e.g.). It would probably be best to wait for this release before merging this and releasing a new tfopgen.

File: tfopgen/create_op.py
Comment (sjperkins) [inline tfopgen/create_op.py:11]: Move `op_libs, op_devices_comps, op_abi` to a newline.

File: README.rst
Comment (sjperkins) [inline README.rst:84]: I think if we start mixing the Makefile options with the operator options, the yaml will start to be cluttered.

How about a `makefile` section to distinguish it from the rest of the yaml?

```yaml
makefile:
    use_old_abi: yes
    libraries:
       - "m"
```

File: README.rst
Comment (sjperkins) [inline README.rst:78]: Rename `devices_compatibilities` to `devices`

File: README.rst
Comment (sjperkins) [inline README.rst:53]: `device compatibilities` to `devices`
`optional external libraries` to `makefile` options

File: README.rst
Comment (sjperkins) [inline README.rst:158]: `Source code will be generated for the requested devices`

File: README.rst
Comment (sjperkins) [inline README.rst:81]: `extra_library_deps` to `libraries`

File: tfopgen/examples/get_slide_patch.yml
Comment (sjperkins) [inline tfopgen/examples/get_slide_patch.yml:19]: Update with makefile section

File: tfopgen/templates/test_source.j2
Comment (sjperkins) [inline tfopgen/templates/test_source.j2:27]: What does this do?

File: README.rst
Comment (sjperkins) [inline README.rst:178]: Change to:

```
Note that gcc version 5 and above uses a new C++ ABI.
However, the official tensorflow binary pip packages are built with gcc 4
which uses an older ABI. Set the option `use_old_abi: yes` to make
the library compatible with the older ABI.

File: tfopgen/templates/Makefile.j2
Comment (sjperkins) [inline tfopgen/templates/Makefile.j2:31]: One possible way of handling the the presence or absence of the `-ltensorflow_framework`   library would be to add `tensorflow_version` to the YAML.

**Review by sjperkins — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
